### ISSUE TITLE
BatchGet with two separate keys returns duplicate

### ISSUE THREAD

**Issue by cody1024d** [guregu/dynamo#67]
@guregu So I'm still trying to track down the cause of this, but I have the following code:

```go
t := db.db.Table("Location")
	k := make([]dynamo.Keyed, len(keys))
	for i, key := range keys {
		k[i] = dynamo.Keys{key, nil}
	}
	var ls []Location
	err := t.Batch("Id").Get(k...).All(&ls)
	retVal := make([]LocationDataResult, len(keys))
	for i := range keys {
		if i < len(ls) {
			retVal[i] = LocationDataResult{&ls[i], err}
		} else {
			retVal[i] = LocationDataResult{Error:err}
			//TODO--If error is nil, we say the object wasn't found
		}
	}
	return retVal
```

And when I pass in 2 keys, where one can be "found" and the other can't (as in, there's a single object in my DB table, and only one of those keys corresponds to it), I receive back 2 copies of the one "valid" object.  I would expect to get back only one result?

I do see that the Keys are created correctly, and are getting appended onto the BatchGet correctly.  That's where I'm at right now, and will continue to investigate further.

**Comment by guregu** [comment]
Try looking at your results here (`ls`) and see if you actually get the same value twice. I have a feeling the problem is with the loop near the end.

**Comment by cody1024d** [comment]
@guregu Yeah I can confirm, when inspecting the ls array itself, there are two of the same values.

Let me know what I can do to help you track this down.

**Comment by guregu** [comment]
Thanks for help. I'll try to reproduce this. If you stick a bunch of logging here https://github.com/guregu/dynamo/blob/1eac96337922d24cb810c5d647d761d25fb01554/batchget.go#L198 and here https://github.com/guregu/dynamo/blob/1eac96337922d24cb810c5d647d761d25fb01554/batchget.go#L248 we might figure it out.

**Comment by guregu** [comment]
Alright, I've done some investigation. I can't reproduce the duplicate bug but I found some other issues. Maybe this fix will help you, try updating. 454c8fd75e4cb650ddfdc8ebcabd28c2b220dcab

**Comment by cody1024d** [comment]
@guregu That seemingly fixed it!  I'm going to continue running my tests to see if it's fully resolved, but on initial runs I'm now only getting back the one result I'd expect.

I'll close this, and reopen it if I find any additional issues.  Thanks!

**Comment by guregu** [comment]
Thanks a lot for the detailed report. There isn't anything particularly asynchronous about the batch get request itself. I'm thinking this is probably some kind of logic error in batch get where the same result gets re-used... maybe the index  in the batch iterator doesn't  get updated properly... hmm...

**Comment by cody1024d** [comment]
No problem.  If there's anything I can step through on the debugger, or put some additional log lines in, let me know!

**Comment by guregu** [comment]
Aha! I've finally managed to reproduce. Should be able to fix soon. Your bug reports were very helpful.

**Comment by guregu** [comment]
OK, I think I've finally fixed it (in c2e53baaac9af29fd95101c6b0a883d091cddb5c). The issue was that the batch get iterator was conflating successful unmarshals with the amount of processed keys. I've changed it to look at the amount of keys actually processed, even if they were missing. I was able to reproduce it in a batch where I requested a missing key and then an existing key. Woohoo! Thank you so much for the cooperation. I owe you a beer 🍻
